core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -858,7 +859,7 @@ class Partition(val topicPartition: TopicPartition,
     // No need to calculate low watermark if there is no delayed DeleteRecordsRequest
     val oldLeaderLW = if (delayedOperations.numDelayedDelete > 0) lowWatermarkIfLeader else -1L
     val prevFollowerEndOffset = replica.stateSnapshot.logEndOffset
-    replica.updateFetchState(
+    replica.updateFetchStateOrThrow(
```

Comment:
I don't think calling `updateFetchState` without holding the leaderIsrUpdate read lock is guaranteed to be safe.

One case I am thinking of is that in `makeLeader` we call:
```
        remoteReplicas.foreach { replica =>
          replica.resetReplicaState(
            currentTimeMs = currentTimeMs,
            leaderEndOffset = leaderEpochStartOffset,
            isNewLeader = isNewLeader,
            isFollowerInSync = partitionState.isr.contains(replica.brokerId)
          )
        }
```
which means that this fetch state update outside of the lock can conflict with the leader epoch bump resetting the replica state. I think acquiring the read lock (and re-validating the leader epoch) potentially simplifies the logic in the PR as well.

Reply:
Thanks for the advice, changing to use the read lock.